### DISCUSSION TITLE
Storing streams offsets using `Consumer#store(long)` is not working (silently)

### DISCUSSION THREAD

**Discussion by aaltihami** [rabbitmq/rabbitmq-stream-java-client#78] [Q&A] [answered]
As per the blog [https://blog.rabbitmq.com/posts/2021/09/rabbitmq-streams-offset-tracking/](url)
  
**" 
The application code handles the tracking explicitly with `Context#storeOffset()` . 
The` Consumer#store(long)` method is another possibility to store an offset 
"** 

The `Consumer#store(long)`  method is supposed to change the offset, However it doesn't store the new offset and there is no exception thrown.

```
environment
	.consumerBuilder()
	.stream(stream)
	.name("consumer_" + stream)
	.manualTrackingStrategy()
	.checkInterval(Duration.ofSeconds(2))
	.builder()
	.offset(OffsetSpecification.first())
	.messageHandler((context, message) -> {
		long messageContent = Long.parseLong(new String(message.getBodyAsBinary()));
		System.out.println(" message : " + messageContent);
		if (someCondition()) {
		    // moving the offset is not working 
		    context.consumer().store(OffsetSpecification.next().getOffset());
		}
	})
	.build();
```

**Comment by acogoluegnes** [comment, marked as answer]
The following line is wrong:

```java
context().consumer().store(OffsetSpecification.next().getOffset());
```

It should be:

```
context().consumer().store(context.offset());
```

**Comment by aaltihami** [comment]
In the use case I'm working on I need to skip all of the stream messages if some condition is fulfilled, Is it possible?

**Reply by acogoluegnes** [reply]
What do you mean by "skip the messages"?

If you don't want to process some messages, you can just not process them and still store their offset. This tells the system that you want to start just after this offset if you resume the consumer later.

**Reply by aaltihami** [reply]
Let's say that:

1- I have pushed 1000 message to the stream.

2- The consumer started to consume the messages from the first offset.

3- During handling the message with the offset 19 (for example), The system noticed that there is no need to process the rest of the messages left in the stream and tries to store the offset 999 directly (jumping from offset 19 to 999 skipping all messages in between)


Based on the above scenario, What I'm  trying to do is moving the offset directly to the end of the stream so the handler method is not even invoked, Is such behavior doable?

**Reply by acogoluegnes** [reply]
You cannot "fast-forward" like this, you have to re-create the consumer.

If the number of messages to skip remains small (a few hundreds), it's OK to just skip them in the message handler.

**Reply by aaltihami** [reply]
Can such functionality be implemented "fast-forward"?

It might be helpful where the number of messages needed to be skipped is large.

And if it's doable, Where it should be implemented, rabbitmq server, rabbitmq Java client or both?

**Reply by acogoluegnes** [reply]
It could be a client-side feature, the library would take care of re-creating a subscription at the appropriate offset. This would be transparent to the application code.

I'll create a follow-up issue, I'd like to see the interest in such a feature, as it may not be trivial to implement.

**Reply by acogoluegnes** [reply]
Follow-up issue: https://github.com/rabbitmq/rabbitmq-stream-java-client/issues/80.

**Reply by aaltihami** [reply]
That's great, thank you.

**Comment by acogoluegnes** [comment]
I will convert this issue to a GitHub discussion. Currently GitHub will [automatically close and lock the issue](https://github.com/github/feedback/discussions/3197) even though your question will be transferred and responded to elsewhere. This is to let you know that we do not intend to ignore this but this is how the current GitHub conversion mechanism makes it seem for the users :(

**Reply by aaltihami** [reply]
Great, Thank you.
